transaction/std/src/memo_builder/gift_code_cancellation_memo_builder.rs

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! Defines the Memo Builder for the gift code cancellation memo (0x0202)
+//! specified in MCIP #32
+
+use super::{
+    memo::{GiftCodeCancellationMemo, UnusedMemo},
+    MemoBuilder, ReservedDestination,
+};
+use mc_account_keys::PublicAddress;
+use mc_transaction_core::{Amount, MemoContext, MemoPayload, NewMemoError};
+
+/// There are three possible gift code memo types specified in MCIP #32
+/// | Memo type bytes | Name                                              |
+/// | -----------     | -----------                                       |
+/// |    0x0002       | Gift Code Sender Memo                             |
+/// |    0x0201       | Gift Code Funding Memo                            |
+/// | -->0x0202<--    | Gift Code Cancellation Memo                       |
+/// This memo builder builds a gift code cancellation memo (0x0202). Gift code
+/// cancellation is defined as the sender sending the gift code TxOut at the
+/// gift code subaddress back to their default address prior to it being spent
+/// by the receiver. When that happens a zero valued TxOut is sent to the gift
+/// code sender's change subaddress with a gift code cancellation memo that
+/// stores the index of the TxOut originally sent to the gift code subaddress
+/// when the gift code was funded.
+#[derive(Clone, Debug)]
+pub struct GiftCodeCancellationMemoBuilder {
+    // Index of the gift code TxOut that was originally funded
+    gift_code_tx_out_global_index: Option<u64>,
+    // Whether or not to enable change memos
+    gift_code_change_memo_enabled: bool,
+    // Whether we've already written the change memo
+    wrote_change_memo: bool,
+}
+
+// Create an empty GiftCodeCancellationMemoBuilder
+impl Default for GiftCodeCancellationMemoBuilder {
+    fn default() -> Self {
+        Self {
+            gift_code_tx_out_global_index: None,
+            gift_code_change_memo_enabled: true,
+            wrote_change_memo: false,
+        }
+    }
+}
+
+impl GiftCodeCancellationMemoBuilder {
+    /// Set the index of the gift code TxOut that was cancelled
+    pub fn set_gift_code_tx_out_index(&mut self, tx_out_global_index: u64) {
+        self.gift_code_tx_out_global_index = Some(tx_out_global_index)
+    }
+    /// Clear the index
+    pub fn clear_gift_code_tx_out_index(&mut self) {
+        self.gift_code_tx_out_global_index = None;
+    }
+    /// Enable change memos
+    pub fn enable_gift_code_change_memo(&mut self) {
+        self.gift_code_change_memo_enabled = true;
+    }
+    /// Disable change memos
+    pub fn disable_gift_code_change_memo(&mut self) {
+        self.gift_code_change_memo_enabled = false;
+    }
```

Comment:
This API seems overly verbose. Consider a more 'fluent' style.

The separate enable/disable methods would make updating the builder based on a toggle UI require a conditional that calls one of two methods, vs. passing the value to one method.

```suggestion
    // Allows callers to pass N, Some(N), or None.
    pub fn gift_code_tx_out_index(mut self, tx_out_global_index: impl Into<Option<u64>>) -> Self {
        self.gift_code_tx_out_global_index = tx_out_global_index.into()
        self
    }
    /// Toggle change memos
    pub fn gift_code_change_memo(mut self, enabled: bool) -> Self {
        self.gift_code_change_memo_enabled = enabled;
        self
    }
```

Reply:
It's done this way to be kept consistent with the other memo builders. impl Into is a pattern I'm not super fund of because it doesn't promote the caller thinking about that they're dealing with an option param.